light/odr.go

```diff
@@ -0,0 +1,106 @@
+// Copyright 2015 The go-ethereum Authors
+// This file is part of the go-ethereum library.
+//
+// The go-ethereum library is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Lesser General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// The go-ethereum library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with the go-ethereum library. If not, see <http://www.gnu.org/licenses/>.
+
+// Package light implements on-demand retrieval capable state and chain objects
+// for the Ethereum Light Client.
+package light
+
+import (
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/crypto"
+	"github.com/ethereum/go-ethereum/ethdb"
+	"github.com/ethereum/go-ethereum/rlp"
+	"golang.org/x/net/context"
+)
+
+// OdrBackend is an interface to a backend service that handles odr retrievals
+type OdrBackend interface {
+	Database() ethdb.Database
+	Retrieve(req OdrRequest) error
+}
+
+// OdrRequest is an interface for retrieval requests
+type OdrRequest interface {
```

Comment:
if you wrap the request in a struct, you can DRY up all the GetCtx redundancies.

```go
type OdrBackend interface {
   Database() ethdb.Database
  Retrieve(req *OdrRequestWithCtx)
}
type OdrRequestWithCtx struct {
 ctx   context.Context
 OdrRequest
}

type OdrRequest interface {
  StoreResult(db ethdb.Database)
}
// GetCtx not needs to be defined exactly once not 5 times
func (req *OdrRequestWithCtx) GetCtx() context.Context { return req.ctx }


// this now does not need the ctx field and the GetCtx method to be defined
type TrieRequest struct {
	root  common.Hash
	key   []byte
	proof []rlp.RawValue
}
```

```

But the question is, do you really need the ctx embedded in the request? cant you always just pass it to Retrieve if it was OdrBackend#Retrieve(req OdrRequest, context.Context)
and eliminate the extra wrapping, GetCtx and ctx fields altogether.